File: ui/src/server/gaps.ts
// Capability gaps — the honesty loop's memory. The contract with agents:
// competence first, and when work genuinely can't be done properly (missing
// tools/access, org-specific process the agent would be guessing at), report
// the gap instead of improvising. The contract with humans: no nagging —
// one row per work-shape ever, repeats bump seen_count (frequency is ranking
// signal), and the Studio's Suggested queue is where a gap gets ratified.
//
// ONE notification, on the FIRST sighting of a work-shape, and never again for
// that shape. That is not a softening of "no inbox pings" — it is the same
// promise, kept: the rule was always one-per-shape, and a queue that nothing
// ever announces is a queue nobody opens. The `gap_reported` class defaults to
// in-app (lib/notifications.ts), so the bell learns about a NEW kind of gap and
// no mail leaves the building unless someone asks for it. Repeats — the
// seen_count bumps that make a shape rank — say nothing at all.
import { db } from './db/pg'
import { addNotification } from './notifications'
import { audienceFor, type Authority } from './approvals'

export interface CapabilityGap {
  id: string
  kind: string
  boardId: string | null
  agentModel: string
  missing: string
  needs: string
  exampleTaskId: string | null
  seenCount: number
  status: 'open' | 'dismissed' | 'resolved'
  createdAt: string
  lastSeen: string
}

const ROW = `id, kind, board_id as "boardId", agent_model as "agentModel", missing, needs,
  example_task_id as "exampleTaskId", seen_count as "seenCount", status,
  created_at as "createdAt", last_seen as "lastSeen"`

const slug = (v: string) =>
  v.trim().toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 60) || 'unclassified'

/** Work-shape identity: the board plus the agent's own name for the kind of
 *  work. Same shape reported again — from any agent — lands on the same row. */
const signatureOf = (boardId: string | null, kind: string) => `${boardId ?? 'any'}|${slug(kind)}`

// ── WHAT AN AGENT'S OWN WORDS MAY BE DISCLOSED AS ────────────────────────────
//
// THE RETRY. Refusing a `taskId` does not take the taskId away from the agent —
// it takes away the BINDING that taskId supplied, and leaves the agent holding
// the same free text with a wider default. PROVEN on this branch: an agent
// refused for a ticket re-sent the identical report with `taskId` removed,
// `boardId` resolved null, `signatureOf(null, kind)` is a different work-shape
// so the row counted as first-seen, and `{ by: 'admin', onBoard: null }` pushed
// the free text — verbatim about that board's invoices — to every admin in the
// workspace, including one with no membership of that board.
//
// Deleting the sentence that TAUGHT the retry (it was in the 403 body, and it
// was still in the MCP tool description every agent reads BEFORE it calls)
// removes the instruction and not the capability. So the capability is closed
// here, once, for every agent-raised subject — a capability gap and an agent
// problem are the same shape and must not answer this question twice.
//
// THE RULE. A refusal is REMEMBERED, briefly, against the agent that was
// refused. An unbounded report from an agent with a live refusal is therefore
// not read as an org-wide claim: it is a report about the ticket that agent was
// just told no about, and it is announced with that ticket's board's authority
// — or, when we could not place the ticket at all, with no authority, so every
// admin hears that it exists and none of them is sent the agent's words.
//
// WHY CORRELATION AND NOT REACH. The alternative weighed was to bound an
// unbounded gap by the boards the agent can work — "an agent that can only see
// board A cannot be quoting board B". It fails on the commonest refusal there
// is: a signed-off or archived ticket is refused on a board the agent IS
// allowed on, so its reach still contains that board and the bound is a no-op.
// It is also toothless in practice, because `allow_all_agents` boards are the
// default and make an agent's reach the whole workspace. And it taxes the one
// case that must keep working: a board-scoped agent reporting "I cannot send
// email at all" is making a claim about the WORKSPACE, not about its boards,
// and every admin should get those words. Correlation touches only the agent
// that was just told no, and only for the next half hour.

const REFUSAL_KEY = 'agent_ticket_refusals'
const REFUSAL_TTL_MS = 30 * 60 * 1000

/** A refusal remembered against ONE agent. `boardId` is the board we are
 *  willing to bind that agent's next unbounded report to, and it is set only
 *  when the agent is genuinely allowed on that board (a ticket a person took
 *  off the table). A ticket on a board the agent may not work — or one that
 *  does not exist — leaves it null: binding to a board the agent cannot see
 *  would let it aim its own text at a chosen board's admins, and "an agent that
 *  can only see board A cannot be quoting board B" cuts both ways. */
interface TicketRefusal {
  boardId: string | null
  at: string
}

/** Remember that this agent was just refused a ticket.
 *
 *  Keyed by the AGENT and by nothing else. Keying it by the agent plus the
 *  subject (the gap's `kind` slug, the problem's summary) would be walked past
 *  by an agent that varied one character between the refusal and the retry, and
 *  buys nothing: the only cost of the coarse key is that an agent told no about
 *  a ticket gets the quieter announcement for its next half hour of reports,
 *  and the quieter announcement still reaches every admin.
 *
 *  Never throws. This runs on the way to a 403 and must not turn a refusal into
 *  a 500 — but note that a write that fails FAILS OPEN, which is why the read
 *  side fails closed.
 *
 *  Stale entries are pruned in the same statement, so the blob cannot grow past
 *  the agents seen in the last half hour. The merge is `||` on jsonb for the
 *  reason `approvals.markAnnounced` argues at length: two refusals in flight at
 *  once must not erase each other, and a read-modify-write over a whole blob is
 *  exactly how they would. */
export async function rememberTicketRefusal(agentModel: string, boardId: string | null): Promise<void> {
  const entry = { [agentModel]: { boardId, at: new Date().toISOString() } satisfies TicketRefusal }
  const cutoff = new Date(Date.now() - REFUSAL_TTL_MS).toISOString()
  try {
    const sql = await db()
    await sql`
      insert into app_settings (key, value) values (${REFUSAL_KEY}, ${sql.json(entry as never)}::jsonb)
      on conflict (key) do update set
        value = coalesce((
          select jsonb_object_agg(e.key, e.value)
          from jsonb_each(app_settings.value) as e
          where (e.value ->> 'at') >= ${cutoff}
        ), '{}'::jsonb) || ${sql.json(entry as never)}::jsonb,
        updated_at = now()
    `
  } catch (e) {
    console.error(`[gaps] could not remember the ticket refusal for "${agentModel}":`, e)
  }
}

/** THE authority an agent's own free text may be announced under.
 *
 *  Both agent-raised subjects ask this — `reportGap` below and
 *  `routes/api/agent.problem.ts` — so the answer cannot differ between them,
 *  and neither of them decides it at the call site. The result goes straight to
 *  `audienceFor`, which is the only thing in the product that turns an authority
 *  into people.
 *
 *  · a ticket the agent WAS allowed to name → that ticket's board. Unchanged,
 *    and the honest path.
 *  · no ticket, no live refusal → `{ by: 'admin' }`. The genuinely org-wide
 *    claim, and every admin gets the words. This is the case the whole design
 *    protects: "I cannot send email at all" is not about a board.
 *  · no ticket, a live refusal we could place → that board. The retry lands
 *    exactly where the honest report would have landed, which also means it
 *    shares the honest report's work-shape signature and stops being a new gap.
 *  · no ticket, a live refusal we could NOT place → `{ by: 'nobody' }`, which
 *    `audienceFor` resolves to an empty `content` and every admin in `fact`.
 *    The agent has demonstrated this report is about a ticket, so it is not an
 *    org-wide claim; we cannot name the board, so nobody is sent the words.
 *
 *  FAILS CLOSED. A memo we could not READ is not evidence that there was no
 *  refusal, and a disclosure decision made from state we could not read must
 *  not resolve to the widest option available. So an unreadable memo is
 *  `{ by: 'nobody' }`: every admin still learns the report exists, and the words
 *  are still in the Studio for anyone who opens it. */
export async function agentTextAuthority(agentModel: string, boardId: string | null): Promise<Authority> {
  if (boardId) return { by: 'admin', onBoard: boardId }
  let memo: TicketRefusal | null
  try {
    const sql = await db()
    const rows = (await sql`
      select value -> ${agentModel} as memo from app_settings where key = ${REFUSAL_KEY}
    `) as unknown as Array<{ memo: TicketRefusal | null }>
    memo = rows[0]?.memo ?? null
  } catch (e) {
    console.error(`[gaps] could not read the refusal memo for "${agentModel}" — announcing the fact only:`, e)
    return { by: 'nobody' }
  }
  const at = memo ? Date.parse(memo.at) : NaN
  if (!Number.isFinite(at) || at < Date.now() - REFUSAL_TTL_MS) return { by: 'admin' }
  return memo!.boardId ? { by: 'admin', onBoard: memo!.boardId } : { by: 'nobody' }
}

export async function reportGap(input: {
  agentModel: string
  kind: string
  missing: string
  needs?: string
  boardId?: string | null
  taskId?: string | null
}): Promise<{ id: string; seenCount: number; first: boolean }> {
  const sql = await db()
  // The authority decides the ROW as well as the announcement. A gap the agent
  // could not bind itself, that we placed from a live refusal, gets that board's
  // signature — so it collapses onto the row the honest report would have
  // written instead of counting as a brand-new work-shape, and a retry after a
  // refusal announces nothing at all the second time.
  const authority = await agentTextAuthority(input.agentModel, input.boardId ?? null)
  const boardId = authority.by === 'admin' ? (authority.onBoard ?? null) : null
  const sig = signatureOf(boardId, input.kind)
  const rows = (await sql`
    insert into capability_gaps (signature, kind, board_id, agent_model, missing, needs, example_task_id)
    values (${sig}, ${slug(input.kind)}, ${boardId}, ${input.agentModel},
            ${input.missing.slice(0, 300)}, ${input.needs?.slice(0, 5000) ?? ''}, ${input.taskId ?? null})
    on conflict (signature) do update set
      seen_count = capability_gaps.seen_count + 1,
      last_seen = now(),
      -- a dismissed shape that keeps recurring reopens; resolved stays resolved
      status = case when capability_gaps.status = 'dismissed' then 'open' else capability_gaps.status end,
      example_task_id = coalesce(capability_gaps.example_task_id, excluded.example_task_id)
    returning id, seen_count as "seenCount", (seen_count = 1) as first
  `) as unknown as Array<{ id: string; seenCount: number; first: boolean }>
  const gap = rows[0]!
  if (gap.first) await announceGap({ ...input, authority })
  return gap
}

/** Tell the admins a NEW kind of gap exists — and tell only the ones who can see
 *  the work it quotes what it actually SAYS.
 *
 *  WHO, and how much, is `audienceFor` in server/approvals.ts: the one answer to
 *  "who should be told about this thing" in the product, which the approvals
 *  census, the judge's QA escalation and this line all resolve through.
 *
 *  The authority is `{ by: 'admin', onBoard }`. Admin, because ratifying a gap
 *  is what the Studio does with it — granting a tool, writing a skill, changing
 *  a process — and none of that is a board member's to do. Bounded by the board,
 *  because `missing` and `needs` are the AGENT'S OWN FREE TEXT, typed while
 *  working one board's ticket, and they routinely quote it: the file, the
 *  customer, the ticket's subject. What this replaced selected the admin list in
 *  raw SQL and mailed 160 characters of `missing` and 800 of `needs` to every
 *  admin in the workspace, board membership or not.
 *
 *  So the admins who can see that board get the gap. Every other admin gets the
 *  FACT — a new kind of gap exists and is waiting to be ratified — carrying not
 *  one word the agent typed. A gap with no board is org-wide and every admin
 *  gets all of it — UNLESS the agent was just refused a ticket, which is what
 *  `agentTextAuthority` above decides. This function does not re-decide it: it
 *  is handed an authority and asks the resolver, which is the whole point.
 *
 *  KNOWN RESIDUAL: the Studio's Suggested queue itself (`listGaps` below) is not
 *  board-scoped, so an admin who follows the link can still read the text. That
 *  is a PULL surface an admin chose to open, not a push into their inbox and
 *  their mail; scoping it changes `listGaps`' callers and is owned by a later
 *  round. The link stays because ratifying is the recipient's actual job.
 *
 *  The kind IS the class (`gap_reported`), which `notifyClassOf` accepts
 *  directly, so the "Capability gaps" control in Settings governs this line.
 *  Until this call site existed that control configured nothing at all.
 *
 *  Never throws: an agent reported a gap honestly and the row is already
 *  written. Losing the notification is a bad day; losing the row — or failing
 *  the agent's POST — would teach the fleet that honesty costs it something. */
async function announceGap(input: {
  agentModel: string
  kind: string
  missing: string
  needs?: string
  authority: Authority
}): Promise<void> {
  try {
    const who = await audienceFor(input.authority)
    const placed = input.authority.by === 'admin' && !!input.authority.onBoard
    for (const userId of who.content) {
      await addNotification(userId, {
        kind: 'gap_reported',
        title: `${input.agentModel} hit a capability gap: ${input.missing.slice(0, 160)}`,
        body:
          `Reported while doing ${input.kind} work.` +
          (input.needs?.trim() ? `\n\nWhat it needs: ${input.needs.trim().slice(0, 800)}` : '') +
          `\n\nRatify it in the Studio's Suggested queue, or dismiss it there.`,
        href: '/studio',
      }).catch((e: unknown) => console.error(`[gaps] could not notify ${userId} of a new gap:`, e))
    }
    for (const userId of who.fact) {
      await addNotification(userId, {
        kind: 'gap_reported',
        title: `${input.agentModel} reported a new kind of capability gap`,
        body:
          (placed
            ? 'It was raised while working a board you are not a member of, so what the agent wrote is not repeated here.'
            : 'It was raised against a ticket the agent was refused, so it is not an org-wide report and what the agent wrote is not repeated here.') +
          "\n\nRatify it in the Studio's Suggested queue, or dismiss it there.",
        href: '/studio',
      }).catch((e: unknown) => console.error(`[gaps] could not notify ${userId} that a new gap exists:`, e))
    }
  } catch (e) {
    console.error('[gaps] could not announce a new capability gap:', e)
  }
}

export async function listGaps(status?: string): Promise<CapabilityGap[]> {
  const sql = await db()
  if (status) {
    return (await sql.unsafe(
      `select ${ROW} from capability_gaps where status = $1 order by seen_count desc, last_seen desc limit 100`,
      [status],
    )) as unknown as CapabilityGap[]
  }
  return (await sql.unsafe(`select ${ROW} from capability_gaps order by seen_count desc, last_seen desc limit 100`)) as unknown as CapabilityGap[]
}

export async function setGapStatus(id: string, status: 'open' | 'dismissed' | 'resolved'): Promise<void> {
  const sql = await db()
  await sql`update capability_gaps set status = ${status} where id = ${id}`
}

export async function openGapCount(): Promise<number> {
  const sql = await db()
  const rows = (await sql`select count(*)::int as count from capability_gaps where status = 'open'`) as unknown as Array<{ count: number }>
  return rows[0]?.count ?? 0
}

File: api/src/gaps.rs
// Capability gaps — the honesty loop's memory. The contract with agents:
// competence first, and when work genuinely can't be done properly (missing
// tools/access, org-specific process the agent would be guessing at), report
// the gap instead of improvising. The contract with humans: no nagging —
// one row per work-shape ever, repeats bump seen_count (frequency is ranking
// signal), and the Studio's Suggested queue is where a gap gets ratified.
//
// ONE notification, on the FIRST sighting of a work-shape, and never again for
// that shape. That is not a softening of "no inbox pings" — it is the same
// promise, kept: the rule was always one-per-shape, and a queue that nothing
// ever announces is a queue nobody opens. The `gap_reported` class defaults to
// in-app, so the bell learns about a NEW kind of gap and no mail leaves the
// building unless someone asks for it. Repeats — the seen_count bumps that
// make a shape rank — say nothing at all.
//
// Port of ui/src/server/gaps.ts. The refusal ladder this module guards (the
// THE RETRY argument) is behavior, not comment: `remember_ticket_refusal` and
// `agent_text_authority` carry the correlation rule verbatim.

use serde::Serialize;
use sqlx::PgPool;

use crate::agent_auth::epoch_ms_to_iso;
use crate::agent_writes::{WriteAuthor, guard_agent_fields};
use crate::approvals::audience_for;
use crate::notify::{NotificationInput, NotifyDeps, add_notification};
use crate::runs::define::Authority;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CapabilityGap {
    pub id: String,
    pub kind: String,
    pub board_id: Option<String>,
    pub agent_model: String,
    pub missing: String,
    pub needs: String,
    pub example_task_id: Option<String>,
    pub seen_count: i32,
    pub status: String,
    pub created_at: String,
    pub last_seen: String,
}

const REFUSAL_KEY: &str = "agent_ticket_refusals";
const REFUSAL_TTL_MS: i64 = 30 * 60 * 1000;

/// `slug` — the agent's own name for the kind of work, folded to a signature
/// atom: lowercase, runs of anything else become one dash, no edge dashes, at
/// most 60 chars, and an empty result is still a classifiable something.
pub fn slug(v: &str) -> String {
    let folded: String = v
        .trim()
        .to_lowercase()
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() { c } else { '-' })
        .collect();
    // TS's `[^a-z0-9]+` → '-' collapses runs; the map above emits runs of '-'
    // which this squeeze then collapses to match.
    let mut squeezed = String::with_capacity(folded.len());
    let mut prev_dash = false;
    for c in folded.chars() {
        if c == '-' {
            if !prev_dash {
                squeezed.push('-');
            }
            prev_dash = true;
        } else {
            squeezed.push(c);
            prev_dash = false;
        }
    }
    let trimmed = squeezed.trim_matches('-');
    let taken: String = trimmed.chars().take(60).collect();
    if taken.is_empty() {
        "unclassified".to_string()
    } else {
        taken
    }
}

/// Work-shape identity: the board plus the agent's own name for the kind of
/// work. Same shape reported again — from any agent — lands on the same row.
fn signature_of(board_id: Option<&str>, kind: &str) -> String {
    format!("{}|{}", board_id.unwrap_or("any"), slug(kind))
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// Remember that this agent was just refused a ticket.
///
/// Keyed by the AGENT and by nothing else — varying one character between the
/// refusal and the retry walks past any finer key, and the only cost of the
/// coarse one is a quieter announcement for the agent's next half hour of
/// reports, which still reaches every admin.
///
/// Never throws: this runs on the way to a 403 and must not turn a refusal
/// into a 500 — but a write that fails FAILS OPEN, which is why the read side
/// (`agent_text_authority`) fails closed. Stale entries are pruned in the same
/// statement, and the merge is `||` on jsonb so two refusals in flight at once
/// cannot erase each other.
pub async fn remember_ticket_refusal(pg: &PgPool, agent_model: &str, board_id: Option<&str>) {
    let entry = serde_json::json!({
        agent_model: {
            "boardId": board_id,
            "at": epoch_ms_to_iso(now_ms()),
        }
    });
    let cutoff = epoch_ms_to_iso(now_ms() - REFUSAL_TTL_MS);
    let res = sqlx::query(
        "insert into app_settings (key, value) values ($1, $2::jsonb) \
         on conflict (key) do update set \
           value = coalesce(( \
             select jsonb_object_agg(e.key, e.value) \
             from jsonb_each(app_settings.value) as e \
             where (e.value ->> 'at') >= $3 \
           ), '{}'::jsonb) || $2::jsonb, \
           updated_at = now()",
    )
    .bind(REFUSAL_KEY)
    .bind(&entry)
    .bind(&cutoff)
    .execute(pg)
    .await;
    if let Err(e) = res {
        tracing::error!("[gaps] could not remember the ticket refusal for \"{agent_model}\": {e}");
    }
}

/// THE authority an agent's own free text may be announced under. Both
/// agent-raised subjects ask this — `report_gap` below and the agent-problem
/// route — so the answer cannot differ between them, and neither of them
/// decides it at the call site. The result goes straight to `audience_for`,
/// which is the only thing in the product that turns an authority into people.
///
/// · a ticket the agent WAS allowed to name → that ticket's board.
/// · no ticket, no live refusal → org-wide admin; every admin gets the words.
/// · no ticket, a live refusal we could place → that board.
/// · no ticket, a live refusal we could NOT place → `Nobody`: every admin
///   still learns the report exists, none is sent the agent's words.
///
/// FAILS CLOSED. A memo we could not READ is not evidence that there was no
/// refusal; an unreadable memo is `Nobody`, not the widest option.
pub async fn agent_text_authority(
    pg: &PgPool,
    agent_model: &str,
    board_id: Option<&str>,
) -> Authority {
    if let Some(b) = board_id {
        return Authority::Admin {
            on_board: Some(b.to_string()),
        };
    }
    let memo: Option<serde_json::Value> = match sqlx::query_scalar::<_, Option<serde_json::Value>>(
        "select value -> $2 from app_settings where key = $1",
    )
    .bind(REFUSAL_KEY)
    .bind(agent_model)
    .fetch_optional(pg)
    .await
    {
        // No row at all (no memo has ever been written) is `rows[0]?.memo`
        // → null in TS — a normal read, handled below, not an error path.
        Ok(memo) => memo.flatten(),
        Err(e) => {
            tracing::error!(
                "[gaps] could not read the refusal memo for \"{agent_model}\" — announcing the fact only: {e}"
            );
            return Authority::Nobody;
        }
    };
    let at_ms = memo
        .as_ref()
        .and_then(|m| m.get("at"))
        .and_then(|v| v.as_str())
        .and_then(|s| chrono::DateTime::parse_from_rfc3339(s).ok())
        .map(|t| t.timestamp_millis());
    match at_ms {
        Some(at) if at >= now_ms() - REFUSAL_TTL_MS => match memo
            .as_ref()
            .and_then(|m| m.get("boardId"))
            .and_then(|v| v.as_str())
        {
            Some(b) => Authority::Admin {
                on_board: Some(b.to_string()),
            },
            None => Authority::Nobody,
        },
        // Missing, malformed, or stale — TS's NaN comparison ladder all lands
        // here too.
        _ => Authority::Admin { on_board: None },
    }
}

/// What `report_gap` hands back to the route: the row's id, its frequency,
/// and whether THIS call was the shape's first sighting.
pub struct ReportedGap {
    pub id: String,
    pub seen_count: i32,
    pub first: bool,
}

/// `reportGap` — the write. The agent's `missing`/`needs` go through the one
/// door first (`capability-gap` surface: the likeliest of all the tools to
/// quote a credential, and none of its outputs were scanned before), then the
/// authority decides BOTH the row's board and the announcement's audience —
/// so a retry after a refusal collapses onto the honest report's signature
/// and announces nothing the second time.
pub async fn report_gap(
    deps: &NotifyDeps,
    input: report_gap::GapInput<'_>,
) -> Result<ReportedGap, sqlx::Error> {
    let mut guarded = [
        Some(input.missing.to_string()),
        input.needs.map(|n| n.to_string()),
    ];
    guard_agent_fields(
        &deps.pg,
        "capability-gap",
        WriteAuthor::Agent(input.agent_model),
        &mut guarded,
        None,
    )
    .await;
    let missing = guarded[0]
        .clone()
        .unwrap_or_else(|| input.missing.to_string());
    let needs = guarded[1].clone();

    let authority = agent_text_authority(&deps.pg, input.agent_model, input.board_id).await;
    let board_id = match &authority {
        Authority::Admin { on_board } => on_board.clone(),
        _ => None,
    };
    let sig = signature_of(board_id.as_deref(), input.kind);
    let (id, seen_count, first): (String, i32, bool) = sqlx::query_as(
        "insert into capability_gaps (signature, kind, board_id, agent_model, missing, needs, example_task_id) \
         values ($1, $2, $3::uuid, $4, $5, $6, $7::uuid) \
         on conflict (signature) do update set \
           seen_count = capability_gaps.seen_count + 1, \
           last_seen = now(), \
           status = case when capability_gaps.status = 'dismissed' then 'open' else capability_gaps.status end, \
           example_task_id = coalesce(capability_gaps.example_task_id, excluded.example_task_id) \
         returning id::text, seen_count, (seen_count = 1) as first",
    )
    .bind(&sig)
    .bind(slug(input.kind))
    .bind(&board_id)
    .bind(input.agent_model)
    .bind(char_take(missing.as_str(), 300))
    .bind(match needs.as_deref() {
        Some(n) => char_take(n, 5000).to_string(),
        None => String::new(),
    })
    .bind(input.task_id)
    .fetch_one(&deps.pg)
    .await?;
    if first {
        announce_gap(
            deps,
            announce_gap::GapAnnounce {
                agent_model: input.agent_model,
                kind: input.kind,
                missing: &missing,
                needs: needs.as_deref(),
                authority: &authority,
            },
        )
        .await;
    }
    Ok(ReportedGap {
        id,
        seen_count,
        first,
    })
}

/// TS `.slice(0, n)` — by UTF-16 code units. Byte-identical for the ASCII
/// these fields are validated as prose in, and char-boundary-safe here where
/// a plain `&s[..300]` could panic on a multibyte edge.
fn char_take(s: &str, n: usize) -> &str {
    match s.char_indices().nth(n) {
        Some((i, _)) => &s[..i],
        None => s,
    }
}

pub mod report_gap {
    pub struct GapInput<'a> {
        pub agent_model: &'a str,
        pub kind: &'a str,
        pub missing: &'a str,
        pub needs: Option<&'a str>,
        pub board_id: Option<&'a str>,
        pub task_id: Option<&'a str>,
    }
}

/// Tell the admins a NEW kind of gap exists — and tell only the ones who can
/// see the work it quotes what it actually SAYS. WHO, and how much, is
/// `audience_for`; this function does not re-decide the authority, it is
/// handed one and asks the resolver, which is the whole point.
///
/// Never throws: an agent reported a gap honestly and the row is already
/// written. Losing the notification is a bad day; losing the row — or failing
/// the agent's POST — would teach the fleet that honesty costs it something.
async fn announce_gap(deps: &NotifyDeps, input: announce_gap::GapAnnounce<'_>) {
    use announce_gap::GapAnnounce;
    let GapAnnounce {
        agent_model,
        kind,
        missing,
        needs,
        authority,
    } = input;
    let who = audience_for(&deps.pg, authority).await;
    let placed = matches!(authority, Authority::Admin { on_board: Some(_) });
    let ratify = "\n\nRatify it in the Studio's Suggested queue, or dismiss it there.";
    for user_id in &who.content {
        let body = format!(
            "Reported while doing {kind} work.{}{ratify}",
            match needs.map(str::trim) {
                Some(n) if !n.is_empty() => format!("\n\nWhat it needs: {}", char_take(n, 800)),
                _ => String::new(),
            }
        );
        if let Err(e) = add_notification(
            deps,
            user_id,
            &NotificationInput {
                kind: "gap_reported",
                title: &format!(
                    "{agent_model} hit a capability gap: {}",
                    char_take(missing, 160)
                ),
                body: Some(&body),
                href: Some("/studio"),
            },
        )
        .await
        {
            tracing::error!("[gaps] could not notify {user_id} of a new gap: {e}");
        }
    }
    for user_id in &who.fact {
        let body = format!(
            "{}{ratify}",
            if placed {
                "It was raised while working a board you are not a member of, so what the agent wrote is not repeated here."
            } else {
                "It was raised against a ticket the agent was refused, so it is not an org-wide report and what the agent wrote is not repeated here."
            }
        );
        if let Err(e) = add_notification(
            deps,
            user_id,
            &NotificationInput {
                kind: "gap_reported",
                title: &format!("{agent_model} reported a new kind of capability gap"),
                body: Some(&body),
                href: Some("/studio"),
            },
        )
        .await
        {
            tracing::error!("[gaps] could not notify {user_id} that a new gap exists: {e}");
        }
    }
}

pub mod announce_gap {
    use crate::runs::define::Authority;
    pub struct GapAnnounce<'a> {
        pub agent_model: &'a str,
        pub kind: &'a str,
        pub missing: &'a str,
        pub needs: Option<&'a str>,
        pub authority: &'a Authority,
    }
}

const GAP_COLS: &str = "id::text, kind, board_id::text, agent_model, missing, needs, \
    example_task_id::text, seen_count, status, \
    (trunc(extract(epoch from created_at) * 1000))::bigint as created_ms, \
    (trunc(extract(epoch from last_seen) * 1000))::bigint as seen_ms";

type GapRow = (
    String,
    String,
    Option<String>,
    String,
    String,
    String,
    Option<String>,
    i32,
    String,
    i64,
    i64,
);

impl From<GapRow> for CapabilityGap {
    fn from(r: GapRow) -> Self {
        CapabilityGap {
            id: r.0,
            kind: r.1,
            board_id: r.2,
            agent_model: r.3,
            missing: r.4,
            needs: r.5,
            example_task_id: r.6,
            seen_count: r.7,
            status: r.8,
            created_at: epoch_ms_to_iso(r.9),
            last_seen: epoch_ms_to_iso(r.10),
        }
    }
}

/// `listGaps` — the Studio's Suggested queue, ranked by frequency then
/// recency, capped at 100.
pub async fn list_gaps(
    pg: &PgPool,
    status: Option<&str>,
) -> Result<Vec<CapabilityGap>, sqlx::Error> {
    let rows: Vec<GapRow> = match status {
        Some(s) => {
            let sql = format!(
                "select {GAP_COLS} from capability_gaps where status = $1 \
                 order by seen_count desc, last_seen desc limit 100"
            );
            sqlx::query_as(sqlx::AssertSqlSafe(sql.as_str()))
                .bind(s)
                .fetch_all(pg)
                .await?
        }
        None => {
            let sql = format!(
                "select {GAP_COLS} from capability_gaps \
                 order by seen_count desc, last_seen desc limit 100"
            );
            sqlx::query_as(sqlx::AssertSqlSafe(sql.as_str()))
                .fetch_all(pg)
                .await?
        }
    };
    Ok(rows.into_iter().map(CapabilityGap::from).collect())
}

pub async fn set_gap_status(pg: &PgPool, id: &str, status: &str) -> Result<(), sqlx::Error> {
    sqlx::query("update capability_gaps set status = $1 where id = $2::uuid")
        .bind(status)
        .bind(id)
        .execute(pg)
        .await?;
    Ok(())
}

pub async fn open_gap_count(pg: &PgPool) -> Result<i32, sqlx::Error> {
    let (n,): (i32,) =
        sqlx::query_as("select count(*)::int from capability_gaps where status = 'open'")
            .fetch_one(pg)
            .await?;
    Ok(n)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn slug_matches_ts() {
        // TS: trim, lower, [^a-z0-9]+ → '-', edge dashes off, 60 max, else
        // 'unclassified'.
        assert_eq!(slug("  Invoice Handling!! "), "invoice-handling");
        assert_eq!(slug("---"), "unclassified");
        assert_eq!(slug(""), "unclassified");
        assert_eq!(slug("A--B___C"), "a-b-c");
        assert_eq!(slug(&"x".repeat(80)).len(), 60);
    }

    #[test]
    fn signature_separates_any_board() {
        assert_eq!(signature_of(None, "Email"), "any|email");
        assert_eq!(signature_of(Some("b1"), "Email"), "b1|email");
    }
}
